src/script_opt/ZAM/ZBody.cc

```diff
@@ -28,6 +30,72 @@
 
 namespace zeek::detail {
 
+#ifdef ENABLE_ZAM_PROFILE
+
+static std::vector<std::shared_ptr<ZAMLocInfo>> caller_locs;
+
+static double compute_CPU_prof_overhead() {
+    double start = util::curr_CPU_time();
+    double CPU = 0.0;
+    const int n = 100000;
+    for ( int i = 0; i < n; ++i )
+        CPU = std::max(CPU, util::curr_CPU_time());
+
+    return 2.0 * (CPU - start) / n;
+}
+
+static double compute_mem_prof_overhead() {
+    double start = util::curr_CPU_time();
+    uint64_t m;
+    util::get_memory_usage(&m, nullptr);
+    const int n = 20000;
+    for ( int i = 0; i < n; ++i ) {
+        uint64_t m2;
+        util::get_memory_usage(&m2, nullptr);
+    }
+
+    return 2.0 * (util::curr_CPU_time() - start) / n;
+}
+
+static double CPU_prof_overhead = compute_CPU_prof_overhead();
+static double mem_prof_overhead = compute_mem_prof_overhead();
+
+#define DO_ZAM_PROFILE                                                                                                 \
+    if ( do_profile ) {                                                                                                \
+        double dt = util::curr_CPU_time() - profile_CPU;                                                               \
```

Comment:
I've brought this up quite a while ago and still holding onto it: This should use CPU time of the main/current thread rather than CPU time of the full process (as curr_CPU_time() currently seems to do). Zeek is multi-threaded, so this includes CPU time consumed by logger or other threads running independently and in parallel to script execution. It seems when pinning a `zeek -r` invocation to a single CPU this could skew and jitter things around.

Last time this came up with script profiling you mentioned "logging/caf" activity may be related to script activity and it would make sense to include in the profile (which I'd still challenge :-) ). However, this is now about individual statements and there's logic to to correct for measurement overhead. Seems to me excluding noise/cpu time used by parallel running threads would be more than reasonable for more consistent measurements first.

Reply:
Happy to leave out other threads from this - please just tell me how to get it! (I'm not sure I remember that for script profiling, either ... but leave the resolution of that uncertainty to your superior Git Issue sleuthing skillz :-)

Reply:
This is was the comment on the prior PR:

https://github.com/zeek/zeek/pull/2087#discussion_r866620021

Could try `CLOCK_THREAD_CPUTIME_ID` for `clock_gettime()` on OSX? Maybe it works on newer versions? Or fallback to PROCESS_CPUTIME. The first [stackoverflow hit](https://stackoverflow.com/questions/17372110/clock-thread-cputime-id-on-macosx) indicates there's another way, too.

I'm also wondering that, if you're going that route, the new function should return `uint64_t` nanoseconds value instead of `double`. It could be a tiny bit faster. If you look really really closely, the `double` division in `curr_CPU_time()` may show up somewhere as being expensive.

https://github.com/zeek/zeek/pull/3380/commits/d08e347e5e6606199627d600039786df5bbe4e6d
https://github.com/zeek/zeek/pull/3380/commits/0c19e6fc5443e4ac6fe84ae0ee10f4a11c58bb5c

Reply:
Thanks for the pointer. Is there something analogous for Linux? Or would this be some `#ifdef` goop? (I guess on `DARWIN`, which I see elsewhere)

Reply:
My take from a little looking is that `clock_gettime(CLOCK_THREAD_CPUTIME_ID, ...)` will work on Linux / FreeBSD, but on OSX it may not and either use process time or what is shown in the stackoverflow post. I don't have OSX readily accessible, you might be better positioned.

On Windows with libunistd, even the current one doesn't seem functional - probably okay: https://github.com/robinrowe/libunistd/blob/aa70d7b123d31f39ce1eb30331a4145123ae4f47/unistd/clock_gettime.cpp#L62-L63

Reply:
I opened https://github.com/zeek/zeek/issues/3659 to capture this.